src/miniscript/astelem.rs

```diff
@@ -750,6 +750,7 @@ impl<Pk: MiniscriptKey, Ctx: ScriptContext> Terminal<Pk, Ctx> {
                     .push_opcode(opcodes::all::OP_EQUAL)
             }
             Terminal::Multi(k, ref keys) => {
+                debug_assert!(!Ctx::is_tapctx());
```

Comment:
nit:

this is a pretty big type smell -- if Multi should not be permitted during Ctx tapctx, perhaps one thing that could be done is to give Multi a type tag which is defined as `enum Void{}`for Taproot and `enum Inhabited {By}` for non Taproot, which disables Multi at the type level.

This is -- by all accounts -- an annoying-ish API to work with, but would ensure that it's impossible to use Taproot with Multi which is superior to debug_assert (especially if we want to hanlde user generated code...)

Reply:
We cannot give Multi a type tag without giving `Terminal` a type tag which would infect `Miniscript` and then `Ctx`. We have enough type tags.

Reply:
the type can be an associated type to Ctx, no?

Reply:
> This is -- by all accounts -- an annoying-ish API to work with, but would ensure that it's impossible to use Taproot with Multi which is superior to debug_assert 

We still ensure that wherever we have any user input(1) FromStr 2) TranslatePk 3) compiler output or 4) From decoded script or 5) Miniscript::from_ast()), we pass it through the context checks, thats check the validity of miniscript rules. If you ever create any instance of `Miniscript<Pk: MiniscriptKey, Tap>` with invalid context rules, the library would error.

```
let ms = Miniscript<Pk, Tap>::from_str("multi(..)").unwrap_err();
let ms = Miniscript<Pk, Segwitv0>::from_str("multi(..)").unwrap();
```

The debug_assert here is merely a sanity check, and if this PR(this commit infact) is correctly implemented should never be triggered.

Reply:
I've implemented this approach here:

https://github.com/sanket1729/rust-miniscript/pull/3

I beleive with this approach I also found a bug in this PR w.r.t. Policy Compilations (Multi Could be emitted from a Tap context it seems?).

Reply:
> I believe with this approach I also found a bug in this PR w.r.t. Policy Compilations (Multi Could be emitted from a Tap context it seems?).

No. Every compiler created `Terminal` calls `Ctx::check_local_validity` (which implies global_validity). 
local validity => check if all possible spend paths are valid spends
global validity => Check if any spend path is valid

https://github.com/rust-bitcoin/rust-miniscript/blob/8ad083a2e4f67e32c6b05ca278cf398089d43cc5/src/policy/compiler.rs#L650 
.

One thing worth considering is if we should really bother adding type safety to the `Terminal` enum. It is a fairly internal thing that users should never really have to deal with. This PR already implements the check for the `Miniscript` struct when creating it from `Terminal` (Miniscript::<Pk, Ctx>::from_ast). 

Most users should always use `Descriptor`, and the few advanced users that use `Miniscript` are also safe. So, we are protecting users that interact with `Terminal` directly. Unfortunately, we have to expose it because users might want `match` the `Terminal` object.

Currently, it is possible to create Miniscript directly from members, but we can do the standard fix.
i.e Add a `new()` method to `Miniscript`, make its members private, provide accessor method, add global checks to new(). This will make sure that there is only one way to create `Minsicript` objects and we implement checks at that point. 

I initially strongly ACKed your suggestion, but now I am considering if it is worth the new code diff. Particularly considering that we might want to stick to this and add another `Ctx::gen_multi_a` with new a new associated type for tapscript `multi_a` (multi-sig with CHECKSIGADD).

Reply:
One note is if you don't like gen_multi, it could be replaced with something like...

```rust
trait EnableDisable : TryFrom<ScriptContextError, Error=ScriptContextError>;
```

and then

```rust
impl TryFrom<ScriptContextError> for Enabled {
   fn try_from(v:ScriptContextError) -> Result<Self, ScriptContextError> {
   Ok(Enabled {})
   }
}
impl TryFrom<ScriptContextError> for Disabled {
   fn try_from(v:ScriptContextError) -> Result<Self, ScriptContextError> {
   Err(v)
   }
}
```

And then you never need any special functions, you would just need to do something like:

```rust
Terminal::Multi(n, k, ScriptContextError::MultiDisabled.try_into()?)
```

The try_into would infer if, based on the current active Ctx, we need to fill Terminal with Enabled or Disabled and appropriately Error out at this point as the type would not be constructible.


Then you would not need any `gen_multi_a` or anything...

Reply:
My primary concern we never had the safety guarantee for `Terminal` w.r.t to consensus rules. For example, we don't have guarantees about height/time combination in the `Terminal` struct. Or you could create a `Terminal<bitcoin::PublicKey, BareCtx>` with Multisig (with n >= 4) or Legacy scripts of more than 520 bytes. Terminal need not even be of type `B`. 

All of these are permitted representations of the `Terminal` object, but are unsafe for consensus. Why does `Multi` deserve special detection? As long as the `Miniscript` struct is type-safe, I think we are good.

Reply:
I think it's an understandable perspective, but generally I fall on the side of wanting to encode as much safety as possible in the type system, to the extent it doesn't prevent a usable API / leaks too much complexity. This seemed like a win and a step in that direction at low code complexity, but perhaps not.

Reply:
I tend to agree that restricting `Terminal` would be hard in general (and potentially undesireable as it would turn many types of analysis errors into parsing errors, which for a user could mean less information or ability to recover), especially as it is mostly an internal-only type.

Reply:
My overall feeling is that we should aim to embed as much correctness properties in the type system as we can. Certain things are hard to do (e.g., checking size limits) in the type system, but other things (valid clause definitions) should be easier to do. That includes things like "does this have a multisig".

The largest benefit of this, from a reviewers perspective, is that I know (just by looking at the types!) that the Tap Policy implementation is correct with respect to Multi clauses, whereas before I was unsure since I had to understand the behavior of a bunch of macros in the compilation code of Policy. The types serve to lighten that review burden by making it clear that it would be a compiler error. When making other refactorings in the future, we can be sure that anything that affects the e.g. Policy compiler will not introduce errors of creating clauses that should not be made. And I think that's fantastic!

W.r.t. better or worse errors at parse time v.s. analysis time, I think we can return the appropriate error during Parse (e.g., no multi allowed) and the patch I sent over does that.